keps/sig-testing/5468-invariant-testing/README.md

```diff
@@ -0,0 +1,960 @@
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+First, all invariant **MUST** have clearly documented owners, which will be surfaced
+with the failure results.
+No invariant tests may roll up to the general test owners.
+There is precedent in documenting featuregate and metrics owners.
+
+A shared system will be introduced to the e2e framework to enable this form of
+testing.
+
+All utilities for this will be owned by the SIG Testing Leads and thoroughly
+vetted for reliability.
+
+New invariants will be tested many times before merge, and may be demoted
+at any time by tagging the associated sentinel test `[Flaky]`. Failure to respond
+to flaking invariant tests in a timeline fashion will result in demoting or
+removing them.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+If implemented poorly, this could result in tests flaking in any number of e2e
+test CI jobs that are now running these tests.
+
+We will mitigate this by thoroughly reviewing and testing the invariant checking utilities, 
+and by limiting to only portable, reliable mechanisms.
+
+If not run in many CI jobs, there will be limited benefit to the signal.
+We will aim to generally introduce these as default selected tests.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+[ginkgo], which we use for the e2e tests, does not have a facility for 
+"this test must run after all other tests". It does have ordering, but only
+within a group of tests. Alternatively we can run arbitrary code after the
+suite, but that does not suit our ability to opt-in/out of the tests.
+
+We will introduce a package under `test/e2e` that contains sentinel test(s) which
+do nothing other than register to ginkgo that they have run, if selected.
+
+These tests can be labeled appropriately, allowing the existing test selection
+mechanisms to select / skip them.
+
+We can then use `ginkgo.ReportAfterSuite` to inspect which, if any, of these 
+sentinel invariant tests ran, and run the *actual* test logic, reporting pass/fail.
```

Comment:
Do we want to run the invariant checking after an incomplete E2E test run? If the E2E suite timed out in the middle or got interrupted during a manual run, the sentinel test may not have run yet, thus (with the current approach) skipping the invariant checking. Given the random ordering of tests, it's non-deterministic.

I think this can be argued either way, but it should be documented. I'm leaning towards always running.

If we want to run it also in case of a partial E2E run, then I think we can: instead of checking *after* the run whether the sentinel *did* run, we check *before* the run whether the sentinel is *meant* to run. This works because `ReportAfterSuite` gets called before the actual run with the result of a dry-run.

Reply:
> Let's just document the behavior for the sake of clarity.

The behavior is the same as any other test (best effort, random scheduling, may fail if the run is terminated early or times out), I'm not sure where we should document this. In one of the KEP sections?

Note that in CI no matter how you implement this we always have the case that tests may not run ... we do not guarantee that the entire CI pod will not be terminated. The scheduling aspect is a relatively unimportant detail of "timeouts / early interrupts will lead to incomplete results" which is true even for non-e2e CI.